### ISSUE TITLE
getHarmMetricsSpatial: make temporal filtering optional

### ISSUE THREAD

**Issue by GreatEmerald** [JornDallinga/JornDallinga.github.io#11]
At the moment, the `getHarmMetricsSpatial()` function uses time series cleaning inside the function. However, this is pointless if one already has run the `cloud_filter()` function and thus has a QC mask for each time step. So, allowing to pass a `qc` parameter directly into the function, instead of passing `cfbands`, `thresholds` and `span`, would eliminate this extra needless processing.

**Comment by GreatEmerald** [comment]
Hm, looks like this is not actually doable after all. On the other hand, disabling the filtering part works well.

So then I'm thinking that also in light of collection 1, the workflow should be changed. `cloud_filter()` creates a mask, then something is needed to apply the mask (I don't see any good options so far, it needs to be multi-core, so some sort of foreach %dopar% mask() call), and finally `getHarmMetricsSpatial()` needs to be called with cloud filtering disabled. That way one gets both the actual cleaned files and also the harmonic parameters.

**Comment by JornDallinga** [comment]
hmm, I added the cloud_filter() function to visualize some steps taken within the getHarmMetricsSpatial() function. I was not sure to add this section, but it does illustrate how the cloud masking works. I hoped it to be a example within the tutorial on a small area. But in light of the heavy required processing and time, we might consider removing it from the tutorial to avoid confusion.  But if we can manage your suggestion on creating a getHarmMeticsSpatial and CloudFilter function that complement each other without doing double work, then yes, that woud be ideal!

**Comment by GreatEmerald** [comment]
Yes, indeed, `cloud_filter()` is actually most useful when standalone. Raw data is not very useful, data from `processProbaVbatch()` is better, but even in Collection 1 the cloud masking algorithm misses cloud shadows and clouds on water, and the result from `cloud_filter()` should be the go-to, usable data. There's no reason to have that attached to harmonic metrics. (Also, I know Eline wanted such data as well, but since it's currently very difficult to get that, she gave up for the time being.)

If it could also output cleaned files instead of a mask, that would also be pretty nice for usability. Although probably would take too long to be worth it, since the bands used for detecting outliers is often times not the bands you want cleaned (I want NDVI cleaned, for instance), whereas a mask can be reused for all bands.

In theory one could also combine `processProbaVbatch()` and `cloud_filter()` into a definitive cleaning solution, but also probably not worth it due to processing time needed, as `cloud_filter()` requires imagery that's masked with the PROBA-V QC layer already. Though if it's possible, in one `foreach`/`mcCalc` call, to apply the PROBA-V QC layer, get the timeseries outliers, and filter also those out, to output the clean data, it might be faster than taking those steps individually.

**Comment by GreatEmerald** [comment]
I looked into possible solutions for masking, but most of the available ones are single-threaded. So I made this quick script that applies the mask from `cloud_filter()` to NDVI files:

https://github.com/GreatEmerald/master-classification/blob/master/src/optical/apply-timeseries-mask.r

This would perhaps be useful to implement into `cloud_filter()` or in addition to it, or even made generic (since `processProbaVbatch() is based on the same idea).

Processing each layer took 60-170 seconds, so around 100 seconds on average, so the processing time for 208 layers was around 11 minutes. Then it took a while to compress it all into one file (about an hour). It could also have an option of passing filenames directly to each thread, so instead of stacking them all into one large file, one could have individual dates to get a file layout like the original.

**Comment by GreatEmerald** [comment]
I just updated the above file to output multiple files as mentioned above, since I realised that `getHarmMetricsSpatial` expects the bands to be in separate files. That also allows skipping the combine part, which means that the whole process of masking (and compressing the output files) took 16 minutes total, which is real nice.

The result takes twice as much space compared to outputting a single brick, though, as the compressor can't optimise across the whole time series. Still, it's 7.1 GB, compared to 12 GB of PROBA-V QC mask cleaned files.
